server/src/remote.rs

```diff
@@ -82,126 +113,189 @@ impl FlightService for FlightServiceHandler {
+
+        // Transmit the entire result.
+        let output = stream::iter(result_set);
         Ok(Response::new(Box::pin(output)))
     }
 
+    /// Insert data points into a table. The name of the table must be provided
+    /// as the first element of `FlightDescriptor.path` and the schema of the
+    /// data points must match the schema of the table.
     async fn do_put(
         &self,
-        _request: Request<Streaming<FlightData>>,
+        request: Request<Streaming<FlightData>>,
     ) -> Result<Response<Self::DoPutStream>, Status> {
-        Err(Status::unimplemented("Not yet implemented"))
+        let mut request = request.into_inner();
+
+        // Extract the table name.
+        let flight_data = request
+            .next()
+            .await
+            .ok_or_else(|| Status::invalid_argument("Missing FlightData."))??;
+
+        debug_assert_eq!(flight_data.data_body.len(), 0);
+
+        let flight_descriptor = flight_data
+            .flight_descriptor
+            .ok_or_else(|| Status::invalid_argument("Missing FlightDescriptor."))?;
+
+        let table_name = flight_descriptor
+            .path
+            .get(0)
+            .ok_or_else(|| Status::invalid_argument("No table name in FlightDescriptor.path."))?;
+
+        // Extract the schema.
+        let schema_or_status = self.get_table_schema_from_default_catalog(&table_name);
+        let schema = if let Ok(schema) = schema_or_status {
+            info!(
+                "Received a RecordBatch with a schema for the table {}.",
+                table_name
+            );
+            schema
+        } else {
+            error!(
+                "Received a RecordBatch with a schema for the missing table {}.",
+                table_name
+            );
+            schema_or_status?
+        };
+
+        // Log that the data points were received.
+        while let Some(flight_data) = request.next().await {
+            let flight_data = flight_data?;
+            debug_assert_eq!(flight_data.flight_descriptor, None);
+            let record_batch = utils::flight_data_to_arrow_batch(
+                &flight_data,
+                schema.clone(),
+                &self.dictionaries_by_id,
+            )
+            .map_err(|error| Status::invalid_argument(format!("{}", error)))?;
+            info!(
+                "Received a RecordBatch with {} data points for the table {}.",
+                record_batch.num_rows(),
+                table_name
+            );
+            // TODO: forward the data points to the StorageEngine.
+        }
+
+        // Confirm the data points were received.
+        Ok(Response::new(Box::pin(stream::empty())))
     }
 
+    /// Not implemented.
     async fn do_exchange(
         &self,
         _request: Request<Streaming<FlightData>>,
     ) -> Result<Response<Self::DoExchangeStream>, Status> {
-        Err(Status::unimplemented("Not yet implemented"))
+        Err(Status::unimplemented("Not implemented"))
```

Comment:
The `Status::unimplemented` messages for `do_exchange`, `do_action`, and `list_actions` are missing a period.

Reply:
Fixed.